nativelib/src/main/scala/scala/scalanative/runtime/Arrays.scala

```diff
@@ -31,381 +36,698 @@ sealed abstract class Array[T]
+  
+// ###sourceLocation(file: "/home/francois/proyectos/oss/scala-native-fbd/scala-native/nativelib/src/main/scala/scala/scalanative/runtime/Arrays.scala.gyb", line: 103)
+
+  val arrayDoubleTypeId = typeId (typeof[scalanative.runtime.DoubleArray])
+  
+// ###sourceLocation(file: "/home/francois/proyectos/oss/scala-native-fbd/scala-native/nativelib/src/main/scala/scala/scalanative/runtime/Arrays.scala.gyb", line: 107)
+    
+  /** 
+    *
+    */      
+  private def typeId (ptr : Ptr[Type]): Int = {
+    (!ptr).id    
+  }
+  
+  /** 
+    *
+    */      
+  private def instanceTypeId (any : AnyRef): Int = {
+    typeId(runtime.getType(any))
+  }
+  
+  /*
+   * validate boundaries and check overlap
+   *
+   * return:
+   *  == 0: no overlap
+   *  != 0: overlap, the sign indicates if forward or backward copy is needed   
+   */   
+  private[runtime] def validateBoundaries (from: Array[_], fromPos: Int, to: Array[_], toPos: Int, len: Int): Int = {
+    if (len < 0)
+      throw new IndexOutOfBoundsException("length is negative")
+
+    if (fromPos < 0 || fromPos + len > from.length)
+      throw new IndexOutOfBoundsException(fromPos.toString)
+
+    if (toPos < 0 || toPos + len > to.length)
+      throw new IndexOutOfBoundsException(toPos.toString)
+    
+    if (from == to) {
+      if (fromPos < toPos && fromPos + len > toPos)
+        -1  // backward copy
+      else if (fromPos > toPos && toPos + len > fromPos)
+        1 // forward copy
+      else
+        0
     }
-
-  protected override def clone(): ObjectArray = {
-    val newarr = ObjectArray.alloc(length)
-    ObjectArray.copy(this, 0, newarr, 0, length)
-    newarr
+    else
+      0
   }
+     
+  /*
+   * only used by ObjectArray
+   */   
+  def alloc(length: Int, arrinfo:  Ptr[Type], sizeofElement: CSize): Ptr[_] = {
+    val arrsize = sizeof[ArrayHeader] + sizeofElement * length
+    val arr = runtime.alloc(arrinfo, arrsize)
+    // set the length
+    !(arr.cast[Ptr[Byte]] + sizeof[Ptr[_]]).cast[Ptr[Int]] = length    
+    arr        
+  } 
 }
 
-object ObjectArray {
-  def copy(from: ObjectArray, fromPos: Int,
-           to: ObjectArray, toPos: Int, length: Int): Unit = {
-    ???
-  }
+/*
+ * 
+ */
+ 
+abstract class PrimitiveArray[T] extends Array[T] {
+  /** called by subclass::clone.  Allocate memory using GC_malloc_atomic and copy header and data */
+  override protected def helperClone(sizeofElement: CSize): Ptr[_] = {
+    val arrsize = sizeof[ArrayHeader] + sizeofElement * length
+    val arr = runtime.allocAtomic(arrsize)
+    `llvm.memcpy.p0i8.p0i8.i64`(arr.cast[Ptr[Byte]], this.cast[Ptr[Byte]], arrsize, 1, false)
+    arr
+  }  
+}
 
-  def alloc(length: Int): ObjectArray = {
-    val arrty = typeof[ObjectArray]
-    val arrsize = sizeof[ArrayHeader] + sizeof[Object] * length
-    val arr = runtime.alloc(arrty, arrsize)
+/*
+ * 
+ */
+
+object PrimitiveArray {
+  /*
+   * Primitive arrays don't contain pointers 
+   */   
```

Comment:
This one should be a single-line comment right above `allocAtomic` call. Javadoc/scaladoc comments should document what corresponding methods do, not how they do it.